File: binary_trees/find_target/find_target.js
class Node {
    constructor(val) {
        this.val = val;
        this.left = null;
        this.right = null;
    };
}

const a = new Node('a');
const b = new Node('b');
const c = new Node('c');
const d = new Node('d');
const e = new Node('e');
const f = new Node('f');

//      a
//     / \
//    b   c
//   /\   /
//  d  e f
a.left = b;
a.right = c;
b.left = d;
b.right = e;
c.left = f;

function find_target_bfs(root, target) {
    if(!root) return false;
    const queue = [root];

    while(queue.length > 0) {
        let current = queue.shift();
        if(current.val === target) {
            return true;
        }
        if (current.left) {
            queue.push(current.left);
        }
        if (current.right) {
            queue.push(current.right);
        }
        
    }

    return false;
}

function find_target_dfs(root, target) {
    if(!root) return false;
    let stack = [ root ];

    while(stack.length > 0) {
        let current = stack.pop();
        if(current.val === target) {
            return true;
        }
        if(current.right) {
            stack.push(current.right);
        }
        if(current.left) {
            stack.push(current.left);
        }
    }

    return false;
}

function find_target_dfs_recursion(root, target) {
    if(!root) return false;
    let targetFound = false;
    function recursion(root) {
        let current = root;
        if(current.val === target) {
            targetFound = true;
        }
        if(current.left) {
            recursion(current.left);
        }
        if(current.right) {
            recursion(current.right);
        }
    }

    recursion(root)
    return targetFound;
}

function find_target_dfs_recursion_optimized(root, target) {
    if(!root) return false;
    if(root.val === target) return true;
    return find_target_dfs_recursion_optimized(root.left, target) || find_target_dfs_recursion_optimized(root.right, target);
}

console.log(find_target_dfs(a, 'f'));
console.log(find_target_dfs(a, 'g'));

console.log(find_target_dfs_recursion(a, 'f'));
console.log(find_target_dfs_recursion(a, 'g'));

console.log(find_target_dfs_recursion_optimized(a, 'f'));
console.log(find_target_dfs_recursion_optimized(a, 'g'));

console.log(find_target_bfs(a, 'f'));
console.log(find_target_bfs(a, 'g'));

File: binary_trees/find_target/find_target.py
from typing import Optional, List
class Node:
    def __init__(self, val):
        self.val = val
        self.left = None
        self.right = None

    def __str__(self):
        return f"value: {self.val}, left: {self.left}, right: {self.right}"

#      a
#     / \
#    b   c
#   /\   /
#  d  e f
a = Node('a')
b = Node('b')
c = Node('c')
d = Node('d')
e = Node('e')
f = Node('f')

a.left = b
a.right = c
b.left = d
b.right = e
c.left = f

def find_target_dfs(root: Optional[Node] | None, target: str) -> bool:
    if root == None:
        print("Root is None: ")
        return None
    
    stack = [root]

    while(len(stack) > 0):
        current = stack.pop()

        if current.val == target:
            return True
        if current.left != None:
            stack.append(current.left)
        if current.right != None:
            stack.append(current.right)
        
    return False

def find_target_dfs_recursion(root: Optional[Node], target: str) -> bool:
    def recursion(root):
        current = root

        if current == target:
            return True
        if current.right != None:
            recursion(current.right)
        if current.left != None:
            recursion(current.left)
    
    recursion(root)

    return False



def find_target_bfs(root: Optional[Node], target: str) -> bool:
    if root == None:
        print("Root is None: ")
        return None
    
    queue = [root]

    while(len(queue) > 0):
        current = queue.pop(0)

        if current.val == target:
            return True
        if current.left != None:
            queue.append(current.left)
        if current.right != None:
            queue.append(current.right)
    
    return False

print("DFS")
print(find_target_dfs(a, "f"))
print(find_target_dfs(a, "j"))
print(find_target_dfs(None, None))

print("DFS Recursion")
print(find_target_dfs(a, "f"))
print(find_target_dfs(a, "j"))
print(find_target_dfs(None, None))

print("BFS")
print(find_target_bfs(a, "f"))
print(find_target_bfs(a, "j"))
print(find_target_bfs(None, None))
